drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/client_involvement.rb

```diff
@@ -0,0 +1,95 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HmisExternalApis::AcHmis
+  class ClientInvolvement
+    attr_accessor :mci_lookup
+    attr_accessor :end_date
+    attr_accessor :mci_ids
+    attr_accessor :start_date
+    attr_accessor :status_message
+
+    def initialize(params)
+      self.end_date = params[:end_date]
+      self.start_date = params[:start_date]
+      self.mci_ids = params[:mci_ids]
+    end
+
+    def validate_request!
+      message = []
+
+      message << 'mci_ids not provided.' if mci_ids.blank?
+      message << 'start_date not provided.' if start_date.blank?
+      message << 'end_date not provided.' if end_date.blank?
+
+      begin
+        self.start_date = Date.strptime(start_date, '%Y-%m-%d') if start_date.present?
+      rescue Date::Error
+        message << 'start_date was not formatted correctly.'
+      end
+
+      begin
+        self.end_date = Date.strptime(end_date, '%Y-%m-%d') if end_date.present?
+      rescue Date::Error
+        message << 'end_date was not formatted correctly.'
+      end
+
+      self.mci_lookup = HmisExternalApis::ExternalId
+        .where(value: mci_ids)
+        .where(source_type: 'Hmis::Hud::Client')
+        .where(namespace: 'ac_hmis_mci')
+        .pluck(:source_id, :value) # i.e. client ID and mci ID
+        .to_h
+
+      # Is this an error condition or not?
+      # message << 'no clients found' if mci_lookup.blank?
+
+      if message.present?
+        self.status_message = message.join(' ')
+      else
+        self.status_message = 'success'
+      end
+    end
+
+    def ok?
+      status_message == 'success'
+    end
+
+    def to_json(_ = nil)
+      {
+        involvements: involvements,
+        status_message: status_message,
+      }.to_json
+    end
+
+    private
+
+    def involvements
+      raise 'check validity first' if status_message.nil?
+
+      return [] unless ok?
+
+      clients = Hmis::Hud::Client.where(id: mci_lookup.keys)
```

Comment:
@gigxz do we want to limit this to a particular data source?  I don't _think_ we need to, since we're using DB ids, but not sure if we should here or on enrollments.

Reply:
Hmm, sure. I don't see why Clients in other Data Sources would have ExternalIDs in the "ac_hmis_mci" namespace, but it doesn't hurt to do a filter on data source here just to be safe.

@yakloinsteak this should do it:

```
      clients = Hmis::Hud::Client.where(id: mci_lookup.keys, data_source_id: data_source.id)


    # elsewhere, copied from Mci
    def data_source
      @data_source ||= HmisExternalApis::AcHmis.data_source
    end
```

Reply:
I'm a little unclear why my development database has 8 hmis data sources and that method assumes only one, but I'll roll with it. Just pushed this fix.

Reply:
Yeah thats because of a bug in the seed